controllers/cluster_controller.go

```diff
@@ -278,70 +296,219 @@ func (r *ClusterReconciler) upgradeControlPlane(ctx context.Context, cluster *ca
+		machineDeployment.Spec.Template.Spec.InfrastructureRef.Name = newInfrastructureMachineTemplate.GetName()
+		machineDeployment.Spec.Template.Spec.Version = to.StringPtr(expectedK8sVersion)
+	}
+
+	err = r.Client.Update(ctx, machineDeployment)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to update MachineDeployment %q", machineDeployment.Name)
+	}
+
+	return nodesWillBeRolled, nil
+}
+
+// upgradeMachinePools will fetch the MachinePools for the given cluster and
+// try to upgrade k8s and OS, setting the right labels.
+func (r *ClusterReconciler) upgradeMachinePools(ctx context.Context, cluster *capi.Cluster, giantswarmRelease *releaseapiextensions.Release) (bool, error) {
+	clusterMachinePools, err := r.getSortedClusterMachinePools(ctx, cluster)
+	if err != nil {
+		return false, err
+	}
 
-			return nil
+	for _, machinePool := range clusterMachinePools {
+		nodesWillBeRolled, err := r.upgradeMachinePool(ctx, cluster, &machinePool, giantswarmRelease)
+		if err != nil {
+			return false, err
+		}
+
+		if nodesWillBeRolled {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// upgradeMachineDeployments will fetch the MachineDeployments for the given
+// cluster and try to upgrade k8s and OS, setting the right labels.
+func (r *ClusterReconciler) upgradeMachineDeployments(ctx context.Context, cluster *capi.Cluster, giantswarmRelease *releaseapiextensions.Release) (bool, error) {
+	logger := r.Log.WithValues("cluster", cluster.Name)
+
+	logger.Info("Fetching the cluster MachineDeployments")
+	clusterMachineDeployments, err := r.getSortedClusterMachineDeployments(ctx, cluster)
+	if err != nil {
+		return false, err
+	}
+
+	logger.Info("Looping through the MachineDeployments to see if they need to be upgraded", "numberOfMachineDeployments", len(clusterMachineDeployments))
+	for _, machineDeployment := range clusterMachineDeployments {
+		waitBeforeUpgradingNext, err := r.upgradeMachineDeployment(ctx, cluster, &machineDeployment, giantswarmRelease)
+		if err != nil {
+			return false, err
+		}
+
+		if waitBeforeUpgradingNext {
+			return true, nil
 		}
+	}
+
+	logger.Info("Finished looping through the MachineDeployments")
 
-		logger.Info("The MachinePool is up to date", "expectedK8sVersion", expectedK8sVersion, "currentK8sVersion", *machinePool.Spec.Template.Spec.Version, "expectedMachineImage", expectedMachineImage, "currentMachineImage", currentMachineImage, "expectedWatchFilterLabel", expectedCAPIVersion, "currentWatchFilterLabel", machinePool.Labels[CAPIWatchFilterLabel])
+	return false, nil
+}
+
+// upgradeWorkers will upgrade both MachinePools and MachineDeployments.
+// While it's technically possible for a cluster to have both types of node
+// pools, we assume no cluster would have both at the same time.
```

Comment:
This would still work if a cluster had both, right? If so maybe:

```
// While it's unlikely the cluster would have both types of node pools, it's
// still technically possible and so it is supported.
```